### PR TITLE
IMAP update; new library and attachment support

### PR DESCRIPTION
**Pull request by glennzw**
This pull request replaces the antiquated `github.com/mxk/go-imap` library with the actively maintained `github.com/emersion/go-imap`. 

I've also added support for reporting campaign emails as attachments.

### CODE REVIEW

**Comment by jordan-wright** [comment]
🎉 Awesome! I’ll give this a look over ASAP. Thanks @glennzw!

File: imap/imap.go
Comment (jordan-wright) [inline imap/imap.go]: Let's avoid having emojis in the comments if possible 😄 Go totally supports it, but I'd like to keep things easy.

File: imap/imap.go
Comment (jordan-wright) [inline imap/imap.go]: If we're not going to keep this log, let's remove this `else`.

File: imap/monitor.go
Comment (jordan-wright) [inline imap/monitor.go]: This is quite a long line. Can we break this up into multiple operations? Like, creating the string, then having `goPhishRegex.FindAllStringSubmatch(content)` or similar?

**Review by jordan-wright — CHANGES_REQUESTED** [review]
Hey @glennzw!

Had some time to look into this a bit. I'd like to rename `imapclient` to the camelCase `imapClient`. I added some suggestions, but then I realized it'll just be easier for you to find/replace 😄 

I added a few other little comments and questions. I think this is looking good! Here in the next couple of days I'll start actually testing it out and I'm sure I'll have more questions for you.

As a side note, I sure do like PR's that remove more code than they add 😛 Great work!

File: imap/imap.go
Comment (glennzw) [reply imap/imap.go]: Ah just a note to myself to make sure I was selecting the appropriate email parts. Removed.

File: imap/imap.go
Comment (glennzw) [reply imap/imap.go]: 🙃

File: imap/imap.go
Comment (glennzw) [reply imap/imap.go]: Hmm I'm not sure, changed.

**Comment by glennzw** [comment]
Great thanks for the feedback @jordan-wright ! Addressed your issues and pushed changes.

**Comment by glennzw** [comment]
I've also added support for invalid TLS certificates on IMAP servers, but I'll wait until this is merged into master as there are some UI and db changes.

**Comment by glennzw** [comment]
IMAP messages can be identified by either sequence number (SeqNum) or unique id (UID). The code above fetches by sequence number, which is fine, but as they change when the inbox changes I'd like to change to UID, but I'm having trouble with the `FetchUid` function. If anyone has some IMAP experience, or just wants to help me out, here's the issue I filed that I'm stuck on:

https://github.com/emersion/go-imap/issues/353

**Comment by S0larflare** [comment]
I’ll take a look tomorrow at my code as I remember having a similar problem with UID from the emersion IMAP

Sent from my iPhone

On 22 Apr 2020, at 13:57, Glenn Wilkinson <notifications@github.com> wrote:

﻿

IMAP messages can be identified by either sequence number (SeqNum) or unique id (UID). The code above fetches by sequence number, which is fine, but as they change when the inbox changes I'd like to change to UID, but I'm having trouble with the FetchUid function. If anyone has some IMAP experience, or just wants to help me out, here's the issue I filed that I'm stuck on:

emersion/go-imap#353<https://github.com/emersion/go-imap/issues/353>

**Comment by glennzw** [comment]
Extra info; this condition happens with Gmail, but on Outlook the Uid function works fine. Weird. Might have to crack open some socat tunnels and inspect traffic 🙄

**Comment by glennzw** [comment]
> I’ll take a look tomorrow at my code as I remember having a similar problem with UID from the emersion IMAP

I found a post from [2012](https://support.google.com/mail/forum/AAAAK7un8RUj-snr8u0p6s/?hl=en-GB) of someone having a similar issue with GMail, the suggestion being to send a NOOP (no op) command to the server. This has fixed the issue.

```
go func() {
	done <- c.Fetch(seqset, items, messages)
}()

c.Noop()
```

IMAP is a horrible piece of technology.

**Comment by glennzw** [comment]
Hey @jordan-wright; I've decided I'm going to stick with SeqNums over UIDs for now. There's a slight advantage to using UIDs for some other gophish features that I'm working on, but not essential for now.

Would appreciate a PR when you get a chance 🙂

File: imap/imap.go
Comment (jordan-wright) [inline imap/imap.go:133]: I like to defer to package constants where possible. It looks like the package has a [`SeenFlag="\\Seen`](https://godoc.org/github.com/emersion/go-imap#pkg-constants) that we might be able to use here.

Here's what that might look like (please double-check me though 😄)

```suggestion
	criteria.WithoutFlags = []string{imap.SeenFlag}
```

File: imap/imap.go
Comment (jordan-wright) [inline imap/imap.go:139]: When possible, I recommend catching and handling conditions early, since that helps avoid long nested blocks.

Here, we can do something like this:

```go
if len(seqs) == 0 {
    return emails, nil
}
// At this point, we know that we have emails we can handle so we can continue as normal.
// The benefit is that we've saved one level of indentation which is cleaner to read.
seqset := new(imap.SeqSet)
...
```

**Comment by jordan-wright** [comment]
Hey @glennzw, just checking in. I'll be working to review this over the next couple of days. My time for PR reviews is spread pretty thin, so I appreciate you bearing with me 😄 

I'll keep you updated!

File: imap/imap.go
Comment (glennzw) [reply imap/imap.go]: Done.

File: imap/imap.go
Comment (glennzw) [reply imap/imap.go]: Good point, done!

File: imap/monitor.go
Comment (jordan-wright) [inline imap/monitor.go:157]: What do you think about removing this `else` and moving this back up a level?

File: imap/monitor.go
Comment (jordan-wright) [inline imap/monitor.go:167]: Should we just `continue` here and bump the `else` up a level?

File: imap/monitor.go
Comment (jordan-wright) [inline imap/monitor.go:171]: Likewise, should we just `continue` here and bump the `else` up a level?

File: imap/monitor.go
Comment (jordan-wright) [inline imap/monitor.go:205]: What's the use case you see where more than one `rid` can be reported per email? I can't really see that happening, but I could be missing something.

If we don't anticipate that happening, this might be a bit easier to implement.

File: imap/monitor.go
Comment (jordan-wright) [inline imap/monitor.go:213]: Small nit: we can make this a bit cleaner. Something like:

```go
rid := r[len(r)-1]
if _, ok := rids[newrid]; !ok {
    rids[newrid] = 0
}
rids[newrid]++
```

File: imap/monitor.go
Comment (jordan-wright) [inline imap/monitor.go:222]: I think that you might prefer the `filepath.Ext` function to grab the extension and check if it's `.eml`. That would be cleaner than checking the filename length and last 4 characters.

```go
ext := filepath.Ext(a.Filename)
if a.Header.Get("Content-Type") == "message/rfc822" || ext == ".eml" {
...
}
```

File: imap/monitor.go
Comment (jordan-wright) [inline imap/monitor.go:226]: Small nit: `attachmentEmail` typo.

File: imap/monitor.go
Comment (jordan-wright) [inline imap/monitor.go:232]: Since you're repeating this code here, you might consider breaking it out into a function. In fact, I think that `checkRIDs` might actually be a good name for that function since it actually checks for the RIDs in content, and we can rename this one to be `matchEmail` or something - I'm open to suggestions!

File: imap/imap.go
Comment (jordan-wright) [inline imap/imap.go:209]: We can remove this comment

**Review by jordan-wright — CHANGES_REQUESTED** [review]
Hey @glennzw!

I finally had a chance to dig into this a little bit. This is looking great! Doing some testing, things seem to work which is awesome.

The comments below are really just some tweaks that might make the code a little bit cleaner. Let me know if you have any questions!

File: imap/imap.go
Comment (glennzw) [reply imap/imap.go:166]: Yes it was one email from Google. I've removed this while we figure out what's going on.

File: imap/imap.go
Comment (glennzw) [reply imap/imap.go:178]: Ah brilliant, nice addition to your library! Always feels good to remove a chunk of superfluous code.

File: imap/monitor.go
Comment (glennzw) [reply imap/monitor.go:157]: If I understand correctly, remove the `else` and change the error check to have a continue to step over to the next email?

```
if err != nil {
	log.Errorf("Error searching email for rids from user '%s': %s", m.Email.From, err.Error())
	continue
} 
if len(rids) < 1 { .......
```

File: imap/monitor.go
Comment (glennzw) [reply imap/monitor.go:167]: Done.

File: imap/monitor.go
Comment (glennzw) [reply imap/monitor.go:171]: Done.

File: imap/monitor.go
Comment (glennzw) [reply imap/monitor.go:205]: So I was trying to cover as many eventualities as possible - perhaps an email has an rid parameter that's not from gophish (some other developer created links with the same rid parameter name/form), or someone forwards two different campaign emails in one.

Happy to remove and simplify if you'd prefer though. We could change the function to singular `checkRID` and just return the first one it finds.

File: imap/monitor.go
Comment (glennzw) [reply imap/monitor.go:213]: Ah, much neater pattern, thanks!

File: imap/monitor.go
Comment (glennzw) [reply imap/monitor.go:222]: Much better, thanks!

File: imap/monitor.go
Comment (glennzw) [reply imap/monitor.go:226]: Eagle eyes, well spotted 🦅

File: imap/monitor.go
Comment (jordan-wright) [reply imap/monitor.go:157]: Right. It looks to me that this would fire if something broke when checking for the regex in the email content.

This would mean that we can't trust the result of `rids`. At that point, since the rest of the loop isn't applicable, we should just continue on with the next email.

File: imap/monitor.go
Comment (jordan-wright) [reply imap/monitor.go:171]: Another question for you. If `HandleEmailReport` fails and gives an error, should we append it to `reportingFailed`?

File: imap/monitor.go
Comment (glennzw) [reply imap/monitor.go:232]: I did think about this when coding it, and have given it some more thought - it's a little tricky as I'm using a map to maintain the unique list of rids.  I think the easiest solution is to implement the two separate functions as you suggest and pass the rid map so we can update it.

Well, I implemented that, let me know if that's OK or if you have a neater idea :)

Also, I do realise we're not checking attachments of attachments, etc. I can save that for later.

File: imap/monitor.go
Comment (jordan-wright) [inline imap/monitor.go:183]: Right now, this fires inside the `for _, m := range msgs {` loop, but it seems as though that's a bug.

Instead, I think the `for` loop should close before we mark any emails as unread or delete them. Imagine the scenario where I get two emails marked as reported in the same run of the monitor. In that case, the first email will be deleted, then in the next iteration of the loop, Gophish will try to delete both the first and the second emails.

This looks like this:

```
time="2020-06-13T08:34:05-05:00" level=info msg="User 'XXXX' reported email with rid FcHeBrc"
time="2020-06-13T08:34:05-05:00" level=info msg="Deleting 1 campaign emails"
time="2020-06-13T08:34:06-05:00" level=info msg="User 'XXXX' reported email with rid 7eXoql6"
time="2020-06-13T08:34:06-05:00" level=info msg="Deleting 2 campaign emails"
```

File: imap/monitor.go
Comment (jordan-wright) [inline imap/monitor.go:174]: You might consider renaming this to `deleteEmails` or some other name that indicates the emails are scheduled to be deleted. Right now, that's the only reason we have this list.

By doing that, you can drop the `im.DeleteReportedCampaignEmail == true` check later on, since you're already checking it here.

That is, unless we had other plans for the campaign emails besides just deleting them. 😄

File: imap/imap.go
Comment (glennzw) [reply imap/imap.go:166]: So on my first  test attachment after remove the CR cleanup I got an error:

`time="2020-06-13T14:36:15+01:00" level=error msg="malformed MIME header line: \r--000000000000a0ce9605a19d6401"`

Previously I'd seen this on GMail with a Google email, but this attachment was on Outlook.com and a crafted phishing email I had from another user on here. Perhaps we can leave the CR cleanup for now and remove once we find a fix?

File: imap/monitor.go
Comment (glennzw) [reply imap/monitor.go:184]: Oh gosh, sloppy! Fixed.

File: imap/monitor.go
Comment (glennzw) [reply imap/monitor.go]: Other plans? Well I was going to add support to Tweet out emails before being deleted, but fair enough I'll rename ;-)

File: imap/imap.go
Comment (jordan-wright) [reply imap/imap.go:164]: Sure, that's reasonable!

File: imap/imap.go
Comment (jordan-wright) [inline imap/imap.go:189]: Small nit - we can move the `if err != nil` check outside the `if mbox.TLS` conditional to avoid repeating ourselves.

```go
if mbox.TLS {
    imapClient, err = client.DialTLS(mbox.Host, new(tls.Config))
} else {
    imapClient, err = client.Dial(mbox.Host)
}
if err != nil {
    return imapClient, err
}
```

File: imap/monitor.go
Comment (jordan-wright) [inline imap/monitor.go:220]: Super small nit: Can we remove any blank lines between the function definition and the start of the function?

File: imap/monitor.go
Comment (jordan-wright) [inline imap/monitor.go:206]: Super small nit: Can we remove any blank lines between the function definition and the start of the function?

File: imap/monitor.go
Comment (jordan-wright) [inline imap/monitor.go:214]: So I thought a bit more about this, and this is really just being used as a set. To that end, we don't actually need to keep track of the number of times we've seen this rid.

This let's us make things even easier, since the ++ line isn't need. To make things _even simpler_ we can use the fact that in Go, retrieving a value using a key that doesn't exist gives you the zero value for that value's type. You can see an example [here](https://blog.golang.org/maps#TOC_4.).

This means that we can just change `rids` to be a `map[string]bool` and then do something like:

```go
if !rids[newrid] {
    rids[newrid] = true
}
```

which is much easier to read.

As a side note, some people may prefer to use `map[string]struct{}` and do the normal `!ok` check since empty `struct{}`'s take up 0 bytes, but personally I like the readability of using a bool, and we're not storing a lot of items here so the cost is negligible, IMO.

File: imap/monitor.go
Comment (jordan-wright) [inline imap/monitor.go:224]: Small nit: if possible, we should try to have a space between the `//` and the comment (the one below also has this issue)

**Comment by jordan-wright** [comment]
Hey @glennzw! This is looking great!

I've added a few more comments. Also I think there are some merge conflicts that we'll need to resolve before we can let the CI pipeline run which handles testing, making sure that `gofmt` has been run, etc.

**Comment by jordan-wright** [comment]
I will say that one thing I really like as we're working through this is that we're actually steadily removing code, which is always a nice thing to see 😄 This PR will be awesome for Gophish.

File: imap/monitor.go
Comment (glennzw) [reply imap/monitor.go]: I like this approach a lot, done!

**Comment by glennzw** [comment]
Hi @jordan-wright ! Made those little changes, thanks for the review! Yeah I do like removing code rather than adding code :D

**Comment by jordan-wright** [comment]
Hey @glennzw!

This LGTM! All that's left is to resolve the merge conflicts and I'll get this merged in.

Thanks for all the hard work on this! 💚

**Comment by glennzw** [comment]
Great! I hope I did the merge conflict correctly;

```
git fetch upstream
git merge master
git push
```

You're very welcome, this has been a fun project to contribute to! 💚

**Comment by glennzw** [comment]
> Great! I hope I did the merge conflict correctly;
> 
> ```
> git fetch upstream
> git merge master
> git push
> ```

Oh, actually, I don't think this is what you were asking me to do 😬

How do I resolve the conflicts?

**Review by jordan-wright — APPROVED** [review]

**Comment by jordan-wright** [comment]
LGTM! Thanks for the great work and continued patience, @glennzw! 🎊

**Comment by glennzw** [comment]
> LGTM! Thanks for the great work and continued patience, @glennzw! 🎊

Awesome! Thanks for all the detailed PR'ing - I learned a lot 🕺

**Pull request merged**
